django/test/runner.py

```diff
@@ -52,19 +54,109 @@ def printErrorList(self, flavour, errors):
             self.stream.writeln("%s" % sql_debug)
 
 
+class RemoteTextTestRunner(unittest.TextTestRunner):
+
+    def run(self, test):
+        # Mimic RemoteTestRunner.run() -- minus code that outputs results.
+        result = self._makeResult()
+        unittest.registerResult(result)
+        result.failfast = self.failfast
+        result.buffer = self.buffer
+        result.startTestRun()
+        try:
+            test(result)
+        finally:
+            result.stopTestRun()
+
+        # Capture tracebacks in the worker process because it's impossible to
+        # pickle tracebacks in general (they can reference arbitrary objects).
+        result.stream = unittest.runner._WritelnDecorator(StringIO())
+        result.printErrors()
+        errors = result.stream.getvalue().strip()
+
+        # Save aggregate statistics because tests methods may be unpickleable.
+        stats = (
+            result.testsRun,
+            len(result.expectedFailures),
+            len(result.unexpectedSuccesses),
+            len(result.skipped),
+            len(result.failures),
+            len(result.errors),
+        )
+
+        return errors, stats
+
+
+def _run_subsuite(subsuite):
+    return RemoteTextTestRunner().run(subsuite)
+
+
+class ParallelTestSuite(unittest.TestSuite):
+
+    def __init__(self, suite, processes):
+        self.subsuites = partition_suite_by_case(suite)
+        self.processes = processes
+        super(ParallelTestSuite, self).__init__()
+
+    def run(self, result):
+        pool = multiprocessing.Pool(processes=self.processes)
+        test_results = pool.imap_unordered(_run_subsuite, self.subsuites)
+
+        errors = []
+        stats = []
+
+        while True:
+            if result.shouldStop:
+                pool.terminate()
+                break
+
+            try:
+                test_errors, test_stats = test_results.next(timeout=0.1)
+            except multiprocessing.TimeoutError:
+                continue
+            except StopIteration:
+                pool.close()
+                break
+
+            errors.append(test_errors)
+            stats.append(test_stats)
+
+        pool.join()
+
+        # Gross hack to trick result into displaying the right data.
```

Comment:
> You can move these lines into a TextTestResult subclass then use it in RemoteTextTestRunner by overriding the resultclass attribute.

Indeed. I tried not to add too many classes but your proposal would be more clean.

> The multiprocessing code can be moved into RemoteTextTestRunner.run() in my opinion.

I'm not sure I understand. `RemoteTextTestRunner` executes on the workers, not on the master.